### ISSUE TITLE
Ideas for cython bindings

### ISSUE THREAD

**Issue by kwmsmith** [uber/h3-py#41]
xref: uber/h3#220

So I spent some time putting together kwmsmith/h3-cython before realizing that y'all have a `cython` branch.  @isaacbrodsky suggested that we merge.  It remains to be seen if that merge would improve things much, since it looks like your cython branch is further along.  I particularly like the following:

* letting Cython/Python manage the lifecycles of your arrays automatically -- very clean; this is definitely the right way to do it.
* scikit-build + better cmake integration than the janky thing I put together.

Some possible improvements would be the following, and were directions I wanted to take the cython bindings; all up for discussion:
* Decoupling the sets-of-hex-string representation from the arrays-of-hex-int64, and allowing consumers to decide whether they want to use the int64 or string representations -- this has come up in my own uses of h3, and was one of the original reasons I started on the cython bindings.
* Leveraging [cython's `array.array` support](http://docs.cython.org/en/latest/src/tutorial/array.html) (see [here](https://github.com/cython/cython/blob/master/Cython/Includes/cpython/array.pxd) as well) for no-copy integration between C-level arrays-of-hex-int64s and `array.array` objects, which has many advantages and nice integration with Cython / Python.
* Building on previous point, we then have seamless, no-copy integration with NumPy arrays for free via the [PEP 3118 buffer protocol](https://www.python.org/dev/peps/pep-3118/), which `array.array` supports.
* Once we can connect h3 indices to numpy efficiently, we then can easily work within Pandas and the general PyData stack as well as tighter (py)spark integration via [pandas UDFs](https://www.twosigma.com/insights/article/introducing-vectorized-udfs-for-pyspark/) again, for free.

To be clear: none of the above rules out the current sets-of-hex-string representation, it only decouples that representation from the array-of-int64 representation, so end users have the choice / option to suit their usecase.  There are some significant performance / memory improvements that would come with the array-of-int64 representation, as well as the seamless Numpy / Pandas / spark integration mentioned above.

Other things:
* I've started doing some property-based testing of h3 via [hypothesis](https://hypothesis.readthedocs.io/en/latest/), which I've found to be really good at finding all those inconsistent edge cases and subtle bugs.  There's also some hypothesis stuff specifically meant for testing [geojson](https://pypi.org/project/hypothesis-geojson/).

**Comment by kwmsmith** [comment]
> Finish wrapping the full C API.

What remains to be wrapped?  Is the "full C API" as defined [here](https://uber.github.io/h3/#/documentation/api-reference/indexing)?

> Get scikit-build working for Python 2/3 on mac/linux, and figure out how to package it to make it pip installable. (This is the trickiest part for us right now.)

The cmake script I cribbed from scikit-build is working for mac/linux, python 3.6 / 3.7.  I made some simplifications compared to yours though.

> I agree on decoupling to allow users to choose either string or int64 representation. That's something we've also wanted to do.

Great!

> array.array makes sense. Would the idea be to use that inside of the HexMem class instead
of what we're currently doing?

Yes -- something like that.

> No copy integration with NumPy sounds great, too. However, I'm a little fuzzy on that part, in how we would think about memory management when we start exposing it. I tried my best to avoid worrying about that in the original design by only exposing Python objects. :)

You actually are already doing everything you need for memory management -- either the `HexMem` object or the `array.array` object owns the memory buffer.  Numpy would just sit on top of that same memory buffer, and hold a reference to the `array.array` object.  It's easier to show than to describe.  

The biggest question, I think, is: is it worth it to introduce a numpy dependency into h3-py?

> Also agree with the benefits of better integration with the PyData stack. In addition, I've thought we might be able to do something nice with Pandas extension arrays.

Cool.

> One concern is about duplicate hexagons in the array representation

Does the H3 C library make any uniqueness guarantees for some or all of its public API?  If not, then I agree, de-duping at the Python layer is necessary, and sets are the obvious thing to use.  I keep on wishing for an easy-to-use C-level hash-based set -- maybe y'all know of one?

If the C library does guarantee unique hexids, then it will be a pretty simple matter to handle null hexes without copying -- we can just swap nulls with non-nulls at the tail of the array, and then take a view of the array -- this will only update internal pointers and won't require any reallocation or copying.

**Comment by nrabinowitz** [comment]
> Does the H3 C library make any uniqueness guarantees for some or all of its public API? 

The C library should guarantee uniqueness in array output (e.g. from `polyfill` and `uncompact`) but does not de-dupe input. Note that C output may also include multiple `null` values, which consumers are intended to ignore.

**Comment by ajfriend** [comment]
> The C library should guarantee uniqueness in array output (e.g. from polyfill and uncompact) but does not de-dupe input. Note that C output may also include multiple null values, which consumers are intended to ignore.

**Input**: If we're really want to optimize for speed, we could offer a "safe" and "unsafe" version of any function that takes a list/set of hexagons as input. The "safe" version would use `set` to dedupe, and the "unsafe" version would skip that, assuming the user knows what they're doing. Thoughts?

**Output**: It sounds like we'd want every function that outputs a set/list of hexagons to drop `null` values. Do you you all agree with that? I assume we could still do that quickly in C/Cython by shifting the the non-`null` values to the start of the array.

**Comment by isaachier** [comment]
Hi everyone, sorry for dropping my work on this since leaving Uber.

> * `array.array` makes sense. Would the idea be to use that inside of [the HexMem class](https://github.com/isaacbrodsky/h3-py/blob/bfb08884e65d5cfe9c8e98167cdee378dcac4467/h3core.pyx#L119) instead
>   of what we're currently doing?
> * No copy integration with NumPy sounds great, too. However, I'm a little fuzzy on that part, in how we would think about memory management when we start exposing it. I tried my best to avoid worrying about that in the original design by only exposing Python objects. :)

This was one primary reason I did not finish my work on h3 Cython bindings even while at Uber. I could not figure out the memory guarantees for wrapped objects and how the malloc'ed C structures would interact with the Python GC.

> * Using hypothesis also sounds like a great idea!
+1

**Comment by ajfriend** [comment]
> What remains to be wrapped? Is the "full C API" as defined [here](https://uber.github.io/h3/#/documentation/api-reference/indexing)?

I was just saying that my version was missing a few things, like `h3_line`, edges, and having `GeoPolygon` support holes. It's possible that the merging our branches would cover everything. @isaacbrodsky @nrabinowitz: Is the [h3 site](https://uber.github.io/h3/#/documentation/api-reference/indexing) up to date on the API? It seems to be missing `h3_line`.

> The cmake script I cribbed from scikit-build is working for mac/linux, python 3.6 / 3.7. I made some simplifications compared to yours though.

I'm more than happy to defer to you on the scikit-build stuff. @isaachier @isaacbrodsky: What do you think?

> You actually are already doing everything you need for memory management -- either the HexMem object or the array.array object owns the memory buffer. Numpy would just sit on top of that same memory buffer, and hold a reference to the array.array object. It's easier to show than to describe.
> 
> The biggest question, I think, is: is it worth it to introduce a numpy dependency into h3-py?

Interesting. If I'm understanding correctly, we could replace some of the memory management code in `HexMem` if we used `array.array` internally. Is that right? We could maybe even drop the `HexMem` class since its main responsibility is to alloc and dealloc memory. The transformation from int64 to sets of hexes could be factored out to a helper function.

Just wondering: If we just exposed the `array.array` object, would we even need to introduce a `numpy` dependency, or could we leave that up to the user? What does everyone else think about a `numpy` dependency?

(@kwmsmith, sorry, I just realized that I said the same thing as you about swapping nulls in my comment above.)

**Comment by kwmsmith** [comment]
@nrabinowitz 

> The C library should guarantee uniqueness in array output 

Excellent.

> but does not de-dupe input.

Hypothetically, what if I were to send in duplicate values to the core C API -- are we in the land of "undefined behavior," could we get segfaults, or does it just slow things down with duplicated effort?

@ajfriend 

> Input: If we're really want to optimize for speed, we could offer a "safe" and "unsafe" version of any function that takes a list/set of hexagons as input. The "safe" version would use set to dedupe, and the "unsafe" version would skip that, assuming the user knows what they're doing. Thoughts?

How unsafe is the "unsafe" path?  If it's even remotely possible that it could lead to segfaults, then we can't allow it even as a possibility, since Python goes to great lengths to guarantee no segfaults.

> Output: It sounds like we'd want every function that outputs a set/list of hexagons to drop null values. Do you you all agree with that? I assume we could still do that quickly in C/Cython by shifting the the non-null values to the start of the array.

Yes, exactly.

One annoyance is that `array.array` instances make a full copy when slicing.  IOW, there's no way to efficiently take a slice of an array after moving the null hexes to the tail.

That said, numpy *can* take efficient slices (called "views"), so if we allow a numpy dependency (and that's a big decision, for sure) then using numpy arrays would allow us to do everything without extraneous copies.

**Comment by nrabinowitz** [comment]
@ajfriend 
> I was just saying that my version was missing a few things, like h3_line, edges, and having GeoPolygon support holes. It's possible that the merging our branches would cover everything. @isaacbrodsky @nrabinowitz: Is the h3 site up to date on the API? It seems to be missing h3_line.

`h3Line` is under [Traversal](https://uber.github.io/h3/#/documentation/api-reference/traversal)

The "official" API (least in terms of which functions bindings should support) is defined by the output of the `binding-functions` script (`./scripts/binding_functions.sh`). h3-js, h3-java, and h3-py all have [parity tests](https://github.com/uber/h3-py/blob/master/tests/test_parity.py) that should fail when you're missing a public API function in the version of H3 C you're using. But of course this doesn't test for conformity around things like e.g. hole handling in `polyfill`.

@kwmsmith
> Hypothetically, what if I were to send in duplicate values to the core C API -- are we in the land of "undefined behavior," could we get segfaults, or does it just slow things down with duplicated effort?

I think it depends on the function... in `h3SetToMultiPolygon` for example, you could get bad/undefined output (e.g. loops with 1 or 2 points), but you shouldn't segfault. You'd also get undefined results in this case if your input set had multiple resolutions. In `compact`, I think you'd get correct output with more processing time. In `uncompact`, I think that dupes in the input result in dupes in the output, but I'd need to verify (we don't have tests for this case).

We could potentially handle these cases in the C lib, but it would have a performance impact, so we've generally chosen to pass the risk on to consumers so they can make their own decisions about whether they need deduping.

**Comment by ajfriend** [comment]
I'll take a stab at an initial implementation in the next few days. I'll just focus on the API, to make sure we agree on that, so I won't focus on memory or speed efficiency. I'll use NumPy in this first pass, and we can decide if we want to drop it later.

I'm imagining we'll have two parallel modules. Maybe `h3._int_rep` and `h3._str_rep`, for working with `int` and `str` hex representations. The `h3._str_rep` would probably be a light wrapper around `h3._int_rep`.

The functions would look something like:

```
int geo_to_h3(double lat, double lng, int res)
str geo_to_h3(double lat, double lng, int res)
int[:] k_ring(int h3_address, int ring_size)
str[:] k_ring(str h3_address, int ring_size)
```

How does that sound? @kwmsmith, does that sound like the right interface for integrating with NumPy?

**Comment by kwmsmith** [comment]
Regarding two modules: makes sense to me.  How would the end user import and use them?  i.e., what would the module's publicly accessible names be?

Other design decisions:
* For the string module, some options for how to represent things:
  * NumPy array of string objects  -- this is how pandas does strings, for example; probably the least surprising option for ppl, IMO.
  * NumPy array of fixed-length strings -- much more compact than former, but not widely used.
  * Python set of string objects -- conforms to the former H3-py API.  Although we could just have convenience functions that do this conversion from numpy -> set explicitly.

Naming conventions:
  * Would it make sense to have naming conventions so that we always use variable names like `h3_indices` for "arrays of int64s" and `h3_addresses` for "arrays of hex-strings"?  We've started doing something like that for the application of H3 at my work and it's very helpful to keep things straight.

Regarding interfacing with NumPy:

* You're using the typed memoryview syntax for the return type, which is nice and clean at the Cython level, but manifests in Python as a `memoryview` object, which isn't something that ppl expect, and would have to explicitly coerce to a numpy array.  I'd recommend for things like `k_ring()` to leave off the return type and return a dynamically typed python object; the actual type of the object to be returned will be a NumPy array of int64's.  [See here, for example](http://docs.cython.org/en/latest/src/userguide/memoryviews.html#coercion-to-numpy)

* For input arrays, it makes sense to use the sliced memoryview syntax, as that tells Cython to expect something that conforms to the buffer protocol.

So something like:

```
int geo_to_h3(double lat, double lng, int res)
str geo_to_h3(double lat, double lng, int res)
k_ring(int h3_address, int ring_size)  # int64 numpy array version
k_ring(str h3_address, int ring_size)  # str array version
compact(int64_t[:] h3_addresses not None)
compact(str[:] h3_addresses not None)
```

Makes total sense as input types.

**Comment by ajfriend** [comment]
@kwmsmith:

I'm playing around with using `array.array` for the automatic memory management, but when I try something like this:

```python
cdef class HexMem2:
    cdef:
        H3int* ptr
        H3int[:] a

    def __cinit__(self, n):
        self.a = array.array('q', [])
        self.a.resize(n)
        # h3c expects zero'd out memory in many cases
        self.a.zero()

        # maybe make this a managed property eventually...
        self.ptr = <H3int*> self.a.data.as_voidptr

    def __len__(self):
        return len(self.a)
```
I get an error like `Python objects cannot be cast to pointers of primitive types`. Any idea what I'm doing wrong here?

**Comment by ajfriend** [comment]
Ok. I'm starting to get the hang of things here. I've got two new implementations that seem to work:

```python
cdef class HexMem2:
    cdef:
        H3Index* ptr
        H3int[:] a

    def __cinit__(self, n):
        # is array.array Python 2.7 friendly, if we are using int64s?
        ar = array.array('q', []) #weird that we have to use 'q' instead of `H3int`
        array.resize(ar, n)
        array.zero(ar) # h3c expects zero'd out memory in many cases
        self.a = ar

        # maybe make this a managed property... it'll change if we resize `a` i think
        self.ptr = <H3Index*> &self.a[0]

    def __len__(self):
        return len(self.a)

    def hexarray(self):
        return np.asarray(self.a)
```

```python
cdef class HexMem3:
    cdef:
        H3Index* ptr
        array.array a

    def __cinit__(self, n):
        a = array.array('q', [])
        array.resize(a, n)
        array.zero(a)
        self.a = a

        self.ptr = <H3Index*> self.a.data.as_voidptr

    def __len__(self):
        return len(self.a)

    def hexarray(self):
        return np.asarray(self.a)
```

One big question: [It doesn't look like Python 2.7 supports using `array.array` with "long long" types like `'q'`](https://cython.readthedocs.io/en/latest/src/tutorial/array.html#data-fields). I think we'll want to maintain Python 2.7 compatibility. Does that mean we can't use `array.array`?

**Comment by kwmsmith** [comment]
Your implementations look like they should work fine; one thing to verify is whether the `np.asarray()` makes a copy or uses the buffer protocol and points to the same buffer as the underlying `array.array`.  I'd slightly prefer the second implementation since it doesn't have to use typed memoryviews at all, which simplifies things.

Also, where are you thinking of putting the logic to move the zeroed h3index values to the end of the array and slice off the end?  Wherever that is done, I'd recommend doing that as numpy, since numpy will not make a copy when slicing in this case, while `array.array` will make a copy.

Regarding python 2.7: 2.7 [will not be supported as of 2020-01-01](https://pythonclock.org/).  Further, a lot of foundational packages have stated that they will [drop 2.7 support](https://python3statement.org/) on or before that date.  Do you have a business need to support it for the next 8 months or so?  If not, I'd recommend just biting the bullet now and dropping 2.7 support.

**Comment by kwmsmith** [comment]
OK, just got clearance from my employer to work on H3 more directly, and I can carve out some work hours for this purpose.  I started a branch a little while ago over on kwmsmith/h3-py that refactors the design somewhat to get the cmake build in a better place.

What's the best way to collaborate?  I see @isaacbrodsky 's cython branch, and it looks like @ajfriend has one too -- where would be the most up-to-date cython bindings I can rebase from?

**Comment by ajfriend** [comment]
> OK, just got clearance from my employer to work on H3 more directly, and I can carve out some work hours for this purpose. I started a branch a little while ago over on kwmsmith/h3-py that reorganizes the packages / modules somewhat to get the cmake build in a better place.

Awesome! I like the reorganization! I wanted to do something similar, but couldn't figure out how with cmake/skbuild.

> What's the best way to collaborate? I see @isaacbrodsky 's cython branch, and it looks like @ajfriend has one too -- where would be the most up-to-date cython bindings I can rebase from?

The plan was to collaborate on @isaacbrodsky's `cython` branch, but then merge that into `h3-py/master` when we get something stable. Can you make a PR on @isaacbrodsky's branch with your new changes?

I've been working on an implementation that avoids the `HexMem` class, and just uses numpy arrays for automatic memory management. I just put it up at https://github.com/ajfriend/h3-py/blob/work/h3core.pyx . After your PR, I can take my changes and try to reorganize them into proper submodules. It does heavily depend on numpy, so I'm not sure that's what we'll choose to do ultimately, but it does make the implementation pretty clean, I think.

> Do you have a business need to support it for the next 8 months or so? If not, I'd recommend just biting the bullet now and dropping 2.7 support.

I'd love to get rid of python 2.7 as fast as possible, but I think we do still have a need to support Python 2.7 for a while. I'll let others weigh in. @isaacbrodsky @nrabinowitz @karentycoon @TengHu

**Comment by ajfriend** [comment]
It looks like we will want to maintain Python 2.7 compatibility for now.

**Comment by ajfriend** [comment]
I have two implementations that I've been playing with:

https://github.com/ajfriend/h3-py/blob/work/h3core.pyx
https://github.com/ajfriend/h3-py/blob/memview/h3core.pyx

The first does memory management using numpy arrays. This is nice because we get automatic clean up, and it makes it easy to expose a numpy array to the user, because the array already owns the memory. However, the downside is that it ties the whole implementation to numpy.

The second one is an attempt to have the core of the wrapper free of the numpy dependency, and make it more optional to tack on later. This approach manages memory with a `HexMem` class that cleans it up when it gets garbage collected. We can use this class to expose a `set` or `list` of hexes in either `str` or `int64` representation, without any dependence on numpy.

If we do want to expose a numpy array to the user without copying the data, it's a little tricker to set up memory ownership so that numpy frees the memory when all views of it have been garbage collected. It is definitely possible tho, as described in @kwmsmith's book.

I think the second approach is the way to go. If that sounds good, I'd like to merge in @kwmsmith's new module organization. I'm currently having issues getting that working, however (#47). And we'd need to update the install to include numpy header info.

**Comment by kwmsmith** [comment]
Hi -- am slammed for the next couple of days; hope to get to this and unblock the build issues by Friday.

**Comment by ajfriend** [comment]
No worries! Sounds good.

**Comment by ajfriend** [comment]
@kwmsmith have you had any luck with the build?

**Comment by kwmsmith** [comment]
Apologies again for the delay -- plan to look at this soon.  Will update you with progress.

**Comment by ajfriend** [comment]
Hey @kwmsmith, we decided we'll move the Cython development to the cython branch on this repo, instead of on @isaacbrodsky's fork.

Since I think we're currently stuck with some build issues, I wonder if we could benefit from syncing more closely to get over this initial hump and unlock future development. Would you be interested in trying to set up some time to chat and figure out the best next steps? Thanks!

**Comment by kwmsmith** [comment]
> Hey @kwmsmith, we decided we'll move the Cython development to the cython branch on this repo, instead of on @isaacbrodsky's fork.

Sounds good.

> Since I think we're currently stuck with some build issues, I wonder if we could benefit from syncing more closely to get over this initial hump and unlock future development. Would you be interested in trying to set up some time to chat and figure out the best next steps? Thanks!

Yep -- the earliest I can sync is Friday.  My H3 time at work has been diverted to...using H3 at work :)  .  I can be sure to carve out the time this week.  Could we draw up a list of blockers and have them ready to go through?

**Comment by ajfriend** [comment]
PR #51 is the current state of things.

Biggest things on my mind with that PR:

- I'm currently unable to modularize the code, because I can't figure out how to get it to build with multiple `.pyx` files. Getting that to work and solidifying the build would be great.
- I'd also like to sync on the design of using the `HexMem` class as a central module, and building the other interfaces around it (`str` vs `int64` and `numpy.array` vs `set`).

Secondary things:

- looking into getting CI working again
- figure out what's needed for Python 2 support

How about we sync sometime next week? Shall we coordinate by email? You can ping me at `ajfriend@uber.com`.

**Comment by ajfriend** [comment]
Also, @kwmsmith, I'd like your thoughts on #51, and if we should go ahead and merge it as-is, or if we should first fix the build to work with multiple `.pyx` files.

**Comment by ajfriend** [comment]
Closing this in favor of #56
